### ISSUE TITLE
Hosting in a machine which cannot connect to internet.

### ISSUE THREAD

**Issue by sarathkcm** [verdaccio/verdaccio#451]
#### My reason:

I am running Verdaccio in a machine which doesn't have internet connectivity (i.e it is in a private network). 

My question is how do I make public packages that are available in the public npmjs registry tobe available in Verdaccio? I don't mean to make the entire npm registry available, but a few packages that my package.json depends on.

I know I can publish individual packages to Verdaccio, after downloading the packaged tar.gz file from the public npmjs repository.
However the packages in my package.json has dependencies and dev-dependencies, which have their own dependencies which in turn have their dependencies. This goes on until the number of dependencies for a single package can be counted in hundreds. For a package to be successfully installed from Verdaccio, all these dependencies must exist in the Verdaccio server. 

Publishing hundreds of packages is a lot of work. Moreover, many of the packages have publishConfig settings that mandate it to publish only to the public npmjs repository, and some have pre-publish, post-publish scripts which need not be run (or gives errors sometimes) while publishing from a packaged tgz file.

I wanted to know if this is a problem the Verdaccio community shares and if there are any ways this can be addressed. 

####  Config file:

Only changes from default config is
```yaml
listen: 0.0.0.0:4873
#commented uplink: registry.npmjs.org      
```

#### Additional information:

 
 - Container:
     - Installed on a server and accessed from develeopment machine.

**Comment by juanpicado** [comment]
hi @sarathkcm . I've read about this problem several times, your approach you describe is not pretty much practical, it would take ages until you finish and does not make sense because there are ways much easier to address this issue. What I've been by far users basically use a computer with access to the internet and perform an install and then copy the storage (which is a single folder) to the offline machine, assuming there aren't restrictions to copy in such place. 

In other words, verdaccio is just a proxy that caches packages, initially needs internet and then can works offline as long all needed packages are already in the storage.

I hope this helps.

**Comment by sarathkcm** [comment]
> use a computer with access to the internet and perform an install and then copy the storage (which is a single folder) to the offline machine.

@juanpicado The approach you mentioned will work for the first time, however, if a new package has to be made available in Verdaccio later it would be trouble again. I think #418 would deal with that problem. 

Also, it's not an offline machine, the network just blocks access to npmjs and any such registries.

> your approach you describe is not pretty much practical, it would take ages until you finish

For the time being, I made a tool that automates the whole process. Takes about 5-10 minutes to process one package.json file with a moderate number of dependencies. I have been using that for about 10 months without issues. But recently it failed to upload one of the packages because the package.json contained a value it didn't recognize. It was then I realized that the tool might need maintenance everytime a new npm version is released.  Before fixing the tool and continue using it, I wanted to know how the community is solving this issue. However, it seems it is not a common use-case for Verdaccio.

**Comment by juanpicado** [comment]
> I made a tool that automates the whole process
Open source?

**Comment by sarathkcm** [comment]
It is not. I made it at work, therefore I am not at liberty to make that decision. The tool was written in C# and it tries to resolve dependencies and download them. Therefore it is likely to break whenever new npm versions are released because the dependency resolution logic becomes outdated.  

I opened this issue to find out and discuss if there were better alternatives to this approach. If there were fellow users with the same problem I had initial thoughts of creating a cli tool which does the same thing, internally uses npm itself to resolve dependencies, and opensource so that it can evolve through community contributions. I guess I won't be trying that now because my use-case for Verdaccio seems to be a pretty rare one.

**Comment by juanpicado** [comment]
> I guess I won't be trying that now because my use-case for Verdaccio seems to be a pretty rare one.

You are not alone, I heard about it several times in this project and also outside as here as an example.
https://github.com/SharePoint/sp-dev-docs/issues/1235#issuecomment-357187255

**Comment by rkent** [comment]
I've been working on this issue recently. Our application is in a prison software teaching environment, where the students do not have access to the internet, but need to be able to run npm to install packages.

I have a script that I run, see https://github.com/Caspia/npm-populate  (Caution that repo is really designed for our very specific use, so it is more of an example than something others could use as is.). We have a list of modules that we want to be usable, along with all of their dependencies.

There are several problems that you encounter, that npm-populate tries to solve:

1) verdaccio keeps the entire history of releases of a particular module, but only downloads versions that you ask for. npm install though grabs that entire list of available versions and decides which one to choose. Usually it wants the most recent, and things are fine. But sometimes the most recent may not be what got downloaded (for example, if a module is a dependency of another module that asked for a specific version). So after we populate the verdaccio cache, we have a separate operation that cleans the list of available versions to only include those that have been downloaded.

2) Similarly, the :latest tag is specified by verdaccio. If you did not download :latest then you need to change :latest to be whatever is the most recent version that you downloaded.

3) In our case, we also wanted to install dependencies from a project on github. Although verdaccio (or is it npm?) allows you to specify github as a source, it will not download the devDependencies for github sources. So we have to locally clone the github repo, and use a normal npm install.

4) We have never come up with a solution for automatically supporting modules with binary elements. Currently the practice that we use is, online, do an npm install --global-style from an online environment with the same node version, then manually move that module with all of its dependencies into the project.
